src/Decompiler/Typing/ComplexExpressionBuilder.cs

```diff
@@ -101,6 +101,52 @@ public Expression BuildComplex(bool dereferenced)
             return exp;
         }
 
+        /// <summary>
+        /// Return fallback expression. This is something like
+        ///    (char *)exp + offset + index
+        /// It used when we could not build appropriate C-like expression.
+        /// e.g. if data type of expression is (int *) and offset is 2 we
+        /// could not generate array access as size of (int *) is 4 (on x86).
+        /// Returning just expression will cause to lose offset. So best way is
+        /// returning '(char *)expression + offset'
+        /// </summary>
+        private Expression FallbackExpression()
+        {
+            Expression e;
+            if (offset == 0 && index == null)
+            {
+                e = expComplex;
+            }
+            else
+            {
+                DataType dt;
+                if (enclosingPtr != null)
+                    dt = new Pointer(PrimitiveType.Char, enclosingPtr.Size);
+                else
+                    dt = PrimitiveType.CreateWord(expComplex.DataType.Size);
+                e = new Cast(dt, expComplex);
+            }
+            if (offset != 0)
```

Comment:
This logic looks very similar to `CreateArrayIndexExpression` in this same class. Could it be reused?